### PR TITLE
deterministic layer_types buffer registration in multiple models

### PR DESCRIPTION
**Pull request by mowoe**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48162) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48162)

# deterministic layer_types buffer registration in multiple models

Hi,

i encountered a problem where when using Gemma3 during a `trl` `GRPOTrainer` run via `accelerate launch` in DDP mode (4x H200), some of the ranks always produced garbage tokens. Which ranks were affected was always random, so it lead me to believe its soem sort of DDP corruption bug.

The problem in the end was that python randomizes the order in which the result of `list(set(config.layer_types))` appears and it does this with a different initial randomness in each process, i.e. in each DDP rank.

DDP's `_sync_module_states` and `_broadcast_coalesced` synchronizes buffers across ranks based on iteration index rather than individual buffer name, which causes swapped RoPE buffers across these ranks.

When launching the `accelerate launch` process with the env variable `PYTHONHASHSEED=0` set (which uses the same seed for all processes, causing the order to be deterministic) everything works fine.

This fix uses `sorted(set())` instead of the non-deterministic `list(set())` and fixes this issue.

## TL;DR
* `modeling_gemma3.py` (and other models) uses `self.layer_types = list(set(config.layer_types))` before registering RoPE buffers (_inv_freq, _original_inv_freq).
* python randomizes string hash seeds per process by default which means the resulting `set()` order is different across DDP ranks
* During initialization, DDP (`_sync_module_states` / `_broadcast_coalesced`) synchronizes buffers across ddp ranks by set order instead of the buffer name
* this can cause global RoPE (\theta=10^6) and sliding-window RoPE (theta=10^4) frequencies being swapped on ranks where the set order does not match exactly the one in rank 0
* as those buffers are registered with `persistent=False` specified, they wont show up in the models `state_dict()` which causes rank weight hashes to be the same even though the buffers are swapped

### CODE REVIEW

**Review by Rocketknight1 — APPROVED** [review]
Yes, this looks sound, both the bug description and the fix seem good. One request before merging: Can you check if Gemma4 (or any other models) are similarly affected, and if so make the same changes there?

**Comment by mowoe** [comment]
Thanks @Rocketknight1 , you were right. In my first commit i just looked for occurrences of `list(set(config.layer_types))`, but turns out that gemma4 family models have the same issue, the `layer_types` initialization just looked slightly different.
I had a look at other model families, but they seem fine, i couldnt find any "non-deterministic"  `set()` operations on buffer initializations.

**Review by Rocketknight1 — APPROVED** [review]
Yes, looks good now!

**Comment by Rocketknight1** [comment]
run-slow: gemma4, diffusion_gemma

**Comment by PatrickNercessian** [comment]
Spent about 8 hours today with Astra + Opus trying to figure out an insane rabbit hole of an issue, and it finally found the same problem. Wish I had found this sooner, nice work! @mowoe
